pageserver/src/tenant/timeline.rs

```diff
@@ -4885,13 +4895,23 @@ impl Timeline {
             return Err(GcError::TimelineCancelled);
         }
 
-        let (horizon_cutoff, pitr_cutoff, retain_lsns) = {
+        let (horizon_cutoff, pitr_cutoff, retain_lsns, max_lsn_with_valid_lease) = {
             let gc_info = self.gc_info.read().unwrap();
 
             let horizon_cutoff = min(gc_info.cutoffs.horizon, self.get_disk_consistent_lsn());
             let pitr_cutoff = gc_info.cutoffs.pitr;
             let retain_lsns = gc_info.retain_lsns.clone();
-            (horizon_cutoff, pitr_cutoff, retain_lsns)
+
+            // Gets the maximum LSN that holds the valid lease.
+            // Caveat: This value could be stale since we rely on refresh_gc_info to invalidate leases,
+            // so there could be leases invalidated between the refresh and here.
+            let max_lsn_with_valid_lease = gc_info.leases.last_key_value().map(|(lsn, _)| *lsn);
```

Comment:
Current implementation would keep everything above the lease. Would be better if this is handled in the same way as retain_lsn so that we only keep necessary images? I forgot where our discussion last time ends up in.

Reply:
I'm actually trying to handle it the same way as what `retain_lsns` does right now. The way current GC uses `retain_lsns` is that for each layer it loops through the `retain_lsn` in `retain_lsns`, and keeping the layer if `layer.get_lsn_range().start <= retain_lsn`.

In my case, I'm doing `layer.get_lsn_range().start <= maximum_lsn_with_valid_lease` (If a layer's `lsn_range` overlaps with the `maximum_lsn_with_valid_lease`, then all other LSNs with valid leases will do as well).

https://github.com/neondatabase/neon/blob/4f13e039b00d51d1bd252c5027852dadc670cdb7/pageserver/src/tenant/timeline.rs#L5051-L5083

Maybe I miss the part where `retain_lsns` only keep necessary images? One way we could do that is to maybe check the end of the lsn range as well?

Reply:
Okay seems like retain_lsn did not optimize for branches (it keeps everything above the oldest branch). Would be a future optimization.